File: examples/python/m2-tool-runtime/tools/ls.py
"""ls：列出工作区目录内容。"""

from __future__ import annotations

from typing import Any, Mapping

from tool_types import ToolSpec
from workspace import Workspace, WorkspaceError

DEFAULT_LIMIT = 500


class LsTool:
    """按字母序列出目录项；目录名带 / 后缀。"""

    def __init__(self, workspace: Workspace) -> None:
        self.workspace = workspace
        self.spec = ToolSpec(
            name="ls",
            description="List directory contents inside the workspace",
            input_schema={
                "type": "object",
                "properties": {
                    "path": {"type": "string"},
                    "limit": {"type": "integer"},
                },
                "additionalProperties": False,
            },
        )

    def validate_arguments(self, arguments: Mapping[str, Any]) -> None:
        path = arguments.get("path")
        if path is not None and not isinstance(path, str):
            raise ValueError("path must be a string")
        limit = arguments.get("limit")
        if limit is not None and (not isinstance(limit, int) or isinstance(limit, bool) or limit < 1):
            raise ValueError("limit must be a positive integer")

    def execute(self, arguments: Mapping[str, Any]) -> dict[str, Any]:
        path = self.workspace.resolve(arguments.get("path"), must_exist=True)
        if not path.is_dir():
            raise WorkspaceError(f"not a directory: {arguments.get('path')!r}")

        limit = int(arguments.get("limit") or DEFAULT_LIMIT)
        names = sorted(path.iterdir(), key=lambda item: item.name.lower())
        entries: list[str] = []
        for item in names[:limit]:
            label = item.name + ("/" if item.is_dir() else "")
            entries.append(label)

        return {
            "path": self.workspace.relative_to_root(path),
            "entries": entries,
            "truncated": len(names) > limit,
            "count": len(entries),
        }

File: examples/rust/m2-tool-runtime/src/tools/ls.rs
use crate::registry::require_object;
use crate::tool_types::{Tool, ToolError, ToolSpec};
use crate::workspace::Workspace;
use serde_json::{Value, json};

const DEFAULT_LIMIT: usize = 500;

/// 按名称列出 Workspace 中的一层目录项。
pub struct LsTool {
    workspace: Workspace,
}

impl LsTool {
    /// 绑定固定 Workspace。
    pub fn new(workspace: Workspace) -> Self {
        Self { workspace }
    }
}

impl Tool for LsTool {
    fn spec(&self) -> ToolSpec {
        ToolSpec {
            name: "ls".into(),
            description: "List directory contents inside the workspace".into(),
            input_schema: json!({
                "type": "object",
                "properties": {
                    "path": {"type": "string"},
                    "limit": {"type": "integer"}
                },
                "additionalProperties": false
            }),
            strict: false,
        }
    }

    fn validate_arguments(&self, arguments: &Value) -> Result<(), ToolError> {
        let object = require_object(arguments)?;
        if let Some(path) = object.get("path") {
            if !path.is_string() {
                return Err(ToolError::new("path must be a string"));
            }
        }
        if let Some(limit) = object.get("limit") {
            let value = limit
                .as_u64()
                .ok_or_else(|| ToolError::new("limit must be a positive integer"))?;
            if value < 1 {
                return Err(ToolError::new("limit must be a positive integer"));
            }
        }
        Ok(())
    }

    fn execute(&mut self, arguments: &Value) -> Result<Value, ToolError> {
        let object = require_object(arguments)?;
        let path_arg = object.get("path").and_then(Value::as_str);
        let path = self.workspace.resolve(path_arg, true)?;
        if !path.is_dir() {
            return Err(ToolError::new(format!("not a directory: {path_arg:?}")));
        }
        let limit = object
            .get("limit")
            .and_then(Value::as_u64)
            .map(|value| value as usize)
            .unwrap_or(DEFAULT_LIMIT);
        let mut names = std::fs::read_dir(&path)
            .map_err(|error| ToolError::new(error.to_string()))?
            .filter_map(Result::ok)
            .map(|entry| entry.path())
            .collect::<Vec<_>>();
        names.sort_by_key(|path| {
            path.file_name()
                .map(|name| name.to_string_lossy().to_ascii_lowercase())
                .unwrap_or_default()
        });
        let truncated = names.len() > limit;
        let entries = names
            .into_iter()
            .take(limit)
            .map(|path| {
                let mut label = path
                    .file_name()
                    .map(|name| name.to_string_lossy().into_owned())
                    .unwrap_or_default();
                if path.is_dir() {
                    label.push('/');
                }
                Value::String(label)
            })
            .collect::<Vec<_>>();
        Ok(json!({
            "path": self.workspace.relative_to_root(&path)?,
            "entries": entries,
            "truncated": truncated,
            "count": entries.len(),
        }))
    }
}
